### PR TITLE
Pull request to merge with development branch

### PR DESCRIPTION
**Pull request by marcos-medeiros**
Thanks for taking the time to review our work!

### Done work

- [x] Create user model

- [x] Create post model

- [x] Create posts and sessions controllers

- [x] Implement authentication features

### Done by:

- @damiecode 
- @marcos-medeiros

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    SecureRandom.urlsafe_base64
+  end
+
+  def self.digest(token)
+    Digest::SHA1.hexdigest(token.to_s)
+  end
+
+  def authenticated?(remember_token)
+    BCrypt::Password.new(remember_digest).is_password?(remember_token)
+  end
```
Comment (maelfosso) [inline app/models/user.rb]: `remember_digest` ? Check it again please

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  def create
+    user = User.find_by(email: params[:session][:email].downcase)
+    if user.authenticate(params[:session][:password])
+      log_in user
+      user.create_remember_token
+      redirect_to :root
+    else
+      render 'new'
+    end
+  end
```
Comment (maelfosso) [inline app/controllers/sessions_controller.rb]: What happens if `user = User.find_by(...)` returns **nil** ?
Need to check that too

File: app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  validates :password, presence: true, length: { minimum: 6, maximum: 255 }
+  has_many :posts
+
+  def self.new_remember_token
+    SecureRandom.urlsafe_base64
+  end
+
+  def self.digest(token)
+    Digest::SHA1.hexdigest(token.to_s)
+  end
```
Comment (maelfosso) [inline app/models/user.rb]: Do it in one method

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -1,2 +1,41 @@
+# frozen_string_literal: true
+
 class ApplicationController < ActionController::Base
+  helper_method :current_user, :logged_in?, :sign_out
+
+  def log_in(user)
+    session[:user_id] = user.id
+  end
```
Comment (maelfosso) [inline app/controllers/application_controller.rb]: No need to use `session` 
- Create the remember token
- Update the user attribute
- Set the current user

File: app/controllers/application_controller.rb [hunk 5]
```diff
@@ -1,2 +1,41 @@
...
+
+  def remember_cookies(user)
+    user.create_remember_token
+    cookies.permanent.signed[:user_id] = user.id
+    cookies.permanent[:remember_digest] = user.remember_token
+  end
```
Comment (maelfosso) [inline app/controllers/application_controller.rb]: You has never called this method anywhere else.
Why ? Did you create it ?

This method, with some changes, can be use when the user log in.
Read the comment before this one to adapt

File: app/controllers/application_controller.rb [hunk 6]
```diff
@@ -1,2 +1,41 @@
...
+    if (user_id = session[:user_id])
+      @current_user ||= User.find_by(id: user_id)
+    elsif (user_id = cookies.signed[:user_id])
+      user = User.find_by(id: user_id)
+      if user&.authenticated?(cookies[:remember_digest])
+        log_in user
+        @current_user = user
+      end
+    end
+  end
```
Comment (maelfosso) [inline app/controllers/application_controller.rb]: It's too long. And after multiple check, it never enters into the second part.
Just use your remember digest and check whether or not there is an user with that remember digest into the database.

File: app/controllers/posts_controller.rb [hunk 7]
```diff
@@ -0,0 +1,29 @@
+# frozen_string_literal: true
+
+class PostsController < ApplicationController
+  before_action :logged_in?, only: %i[new create]
```
Comment (maelfosso) [inline app/controllers/posts_controller.rb]: I can access to `posts/new` because you don't redirect me if I'm not logged in
Correct that, please

Comment (maelfosso) [reply app/controllers/posts_controller.rb]: Or create another method in `application_controller` where you will redirect me if I'm not logged in and use that method here in `before_action`

File: app/controllers/sessions_controller.rb [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+class SessionsController < ApplicationController
+  def new
+    render 'new'
+  end
+
+  def create
+    user = User.find_by(email: params[:session][:email].downcase)
+    if user.authenticate(params[:session][:password])
+      log_in user
+      user.create_remember_token
```
Comment (maelfosso) [inline app/controllers/sessions_controller.rb]: Those 2 lines can be done in only one

File: app/controllers/application_controller.rb [hunk 9]
```diff
@@ -1,2 +1,32 @@
+# frozen_string_literal: true
+
 class ApplicationController < ActionController::Base
+  helper_method :current_user, :logged_in?, :sign_out
+
+  def log_in(user)
+    remember_token = User.new_remember_token
+    cookies.permanent[:remember_token] = remember_token
+    user.update_attribute(:remember_token, User.digest(remember_token))
```
Comment (maelfosso) [inline app/controllers/application_controller.rb]: Why do you digest it again ?
It's already digest !!

File: app/models/user.rb [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+  validates :password, presence: true, length: { minimum: 6, maximum: 255 }
+  has_many :posts
+
+  def self.new_remember_token
+    Digest::SHA1.hexdigest(SecureRandom.urlsafe_base64)
+  end
+
+  def self.digest(token)
+    Digest::SHA1.hexdigest(token.to_s)
+  end
```
Comment (maelfosso) [inline app/models/user.rb]: Not needed

File: app/controllers/application_controller.rb [hunk 11]
```diff
@@ -1,2 +1,32 @@
...
+  end
+
+  def current_user?
+    @current_user = user
+  end
+
+  def current_user
+    remember_token = User.digest(cookies[:remember_token])
+    @current_user ||= User.find_by(remember_token: remember_token)
+  end
```
Comment (maelfosso) [inline app/controllers/application_controller.rb]: No need to digest it again !!
And it's `cookies.permanent`

File: app/controllers/application_controller.rb [hunk 12]
```diff
@@ -1,2 +1,32 @@
...
+
+  def logged_in?
+    !current_user.nil?
+  end
+
+  def sign_out
+    current_user.update_attribute(:remember_token,
```
Comment (maelfosso) [inline app/controllers/application_controller.rb]: It's not needed

File: app/controllers/application_controller.rb [hunk 13]
```diff
@@ -1,2 +1,31 @@
...
+  end
+
+  def current_user?
+    @current_user = user
+  end
+
+  def current_user
+    user = User.find_by(id: cookies.signed[:user_id])
+    @current_user ||= user&.remember_digest == cookies[:remember_digest] ? user : nil
+  end
```
Comment (samgaco) [inline app/controllers/application_controller.rb]: :+1:

**Pull request merged**